File: PasswordBasedHashing/HashEncryptionService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Security.Cryptography;
using System.Text;
using Microsoft.Extensions.Options;

namespace PasswordBasedHashing
{
    public class HashEncryptionService : IHashEncryption
    {

        private const int SaltSize = 16; // 128 bit
        private const int KeySize = 32; //256 bit

        /// <summary>
        /// <para>Initializer for the service.</para>
        /// </summary>
        /// <param name="options">How many iterations for PBKDF2</param>
        public HashEncryptionService(IOptions<HashingOptions> options)
        {
            Options = options.Value;
        }

        private HashingOptions Options { get; }

        /// <summary>
        /// <para>Given an input string, this method returns an encrypted hash in the format of `{iterations}.{salt}.{hash}`.</para>
        /// <para>Uses the SHA256 hash algorithm and [System.Security.Cryptography.Rfc2898DeriveBytes]'s PBKDF2.</para>
        /// <para>Salt size and key size are hard coded. Iteration number is passed into service initialization options.</para>
        /// </summary>
        /// <param name="input">The input string to be hashed.</param>
        /// <returns>Encrypted hash of the input string</returns>
        public string Hash(string input)
        {
            using Rfc2898DeriveBytes algorithm = new Rfc2898DeriveBytes(
                input,
                SaltSize,
                Options.Iterations,
                HashAlgorithmName.SHA256);
            string key = Convert.ToBase64String(algorithm.GetBytes(KeySize));
            string salt = Convert.ToBase64String(algorithm.Salt);

            return $"{Options.Iterations}.{salt}.{key}";
        }

        /// <summary>
        /// <para>Determines equality for provided hash and input strings.</para>
        /// <para>Hashes the input string, then compares it's bytes to the provided hash's bytes for equality.</para>
        /// </summary>
        /// <param name="hash">A hashed value to compare with the input string.</param>
        /// <param name="input">Assumed original string value of the provided hash.</param>
        /// <returns>A bool indicating equality, and another bool indicating whether the provided hash's iteration count
        /// does not equal the current service's iteration count.</returns>
        public (bool Verified, bool NeedsUpgrade) Check(string hash, string input)
        {
            string[] parts = hash.Split('.', 3);

            if (parts.Length != 3)
            {
                throw new FormatException("Unexpected hash format. " +
                                          "Should be formatted as `{iterations}.{salt}.{hash}`");
            }

            int iterations = Convert.ToInt32(parts[0]);
            byte[] salt = Convert.FromBase64String(parts[1]);
            byte[] key = Convert.FromBase64String(parts[2]);

            bool needsUpgrade = iterations != Options.Iterations;

            using Rfc2898DeriveBytes algorithm = new Rfc2898DeriveBytes(
                input,
                salt,
                iterations,
                HashAlgorithmName.SHA256);
            byte[] keyToCheck = algorithm.GetBytes(KeySize);

            bool verified = keyToCheck.SequenceEqual(key);

            return (verified, needsUpgrade);
        }
    }
}

File: goLang/hashEncryptionService.go
package passwordBasedHashing

import (
	"bytes"
	"crypto/sha256"
	"encoding/base64"
	"fmt"
	"os"
	"strconv"
	"strings"

	"golang.org/x/crypto/pbkdf2"
)

const (
	SaltSize   = 16
	KeySize    = 32
	Iterations = 10000
)

func Hash(input string) string {
	salt := make([]byte, SaltSize)
	key := pbkdf2.Key([]byte(input), salt, Iterations, KeySize, sha256.New)

	saltString := base64.StdEncoding.EncodeToString([]byte(salt))
	keyString := base64.StdEncoding.EncodeToString([]byte(key))

	return fmt.Sprint(Iterations, ".", saltString, ".", keyString) //example output: {10000.a9834ythakg.p1dkfh48bai32}
}

func EqualityCheck(hash string, input string) (Verified bool, NeedsUpgrade bool) {
	parts := strings.Split(hash, ".")
	if len(parts) != 3 {
		//todo
	}
	iterations, err := strconv.Atoi(parts[0])
	if err != nil {
		fmt.Println(err)
		os.Exit(2)
	}
	salt, err := base64.StdEncoding.DecodeString(parts[1])
	if err != nil {
		fmt.Println(err)
		os.Exit(2)
	}
	key, err := base64.StdEncoding.DecodeString(parts[2])
	if err != nil {
		fmt.Println(err)
		os.Exit(2)
	}

	needsUpgrade := iterations == Iterations
	generatedKey := pbkdf2.Key([]byte(input), salt, Iterations, KeySize, sha256.New)
	if bytes.Compare(generatedKey, key) == 0 {
		return true, needsUpgrade
	}
	return false, needsUpgrade
}
